[1.13] [WIP] Extend the Tag System

MajorTuvok opened the pull request:
Follow up of my previous attempt (#5167). (this PR uses great parts of the existing code from ForgeTagCollection (from the previous PR))

### Purpose
As has been discussed, it seems to be necessary to extend Minecrafts Tag System to not only allow tagging of Items,Blocks and Fluids but literally everything. Additionally there needs to be a solution for cross-mod compat (see below) in order to allow for a successful interaction of Modded Tags with each other. As a third goal, this PR attempts to make modders life easier, by providing them with the possibiltiy to setup tagging for their own Registry with only a few lines of code (instead of having to implement all the loading and networking required by themselves). 

### Networking
All Forgeregistries are added to the new ForgeTagManager, allowing to create Tags for quite a few things in Minecraft, whilst at the same time keeping compat with vanilla networking code. 
If connected to an vanilla client, then this will simply ignore the additional data sent to it. If connected to a Vanilla Server, then the Tags will not be updated. A Forge-Forge Connection will as expected sync all Tags, except if the TagCollection reports that it doesn't want to be synced.

### Loading
In general a Tag-Collection is responsible for loading it's Tags and modders can therefore specify the folders to load their Tags from individually (at construction time of the ForgeTagCollection or via sub-classing). Currently this PR makes Forge load it's Tags from:

- [tag-owner-namespace]/tags/[registry-namespace]/[registry-name]
- [tag-owner-namespace]/tags/[registry-name]

The latter location is required to preserve compat with Vanilla Data Packs and is also possible for non-Vanilla Forge-Tags for consistency reasons. The first method is possible and I guess will be preferred by Data Pack authors with multiple mods, as it allows for a better grouping of Registries, though everyone can do as he/she likes. In case of conflicts (I trust modders not to use other mods mod-ids as their registry-names, so this should not happen), no special detection is done - though all the failing Tags should tell the Data Pack author that something is going wrong. The Tags can then simply be moved to a structure mimicing the first option. (For resons of consistency I'd recommend using ForgeTagCollection#getDefaultLoadingLocation to get the loading Parameters (implementing: see below)) 

### Loading - Things of Note
Because I was already fiddeling with Tag-Loading, I also improved Error-handling of failing Tags, creating 2 additional Exceptions in the process. MissingTagsException and MissingEntriesException both extend the more general TagException class and provide more context for printing out much more useful errors than Vanilla does. 
Additionally failing because Entries could not be resolved is post-boned to after all Tags have been parsed, helping people like me who often add more than one file with more than one error not to need to restart the Game what feels like a thousand times.

### Cross-mod compat - Vanilla's Problem
Why do we even need additional cross-mod compat, you might ask yourselves? Vanilla already allows to replace other Tags, so why should we need additional compat?

Consider the following case: Mod A creates some kind of awesome Tag. Mod B now wants to append it's own objects to this Tag and therefore replaces it with it's own Tag, containing the same Items A had, and a few more. Mod C now does the same, not knowing that B added Items... 
This results in one of the 2 Mods hiding the others Tagged Items (which one get's precedence depends on loading order) which obviously isn't wished. Also replacing requires a Tag to be declared in the same namespace as the other Tag, thereby violating the contract that a mod should not add resources under another mod's namespace.

### Cross-mod compat - Suggested Solution
Cross-mod compat is improved by implementing a suggestion, which was made quite a while ago. 
Each and every Tag will now be referenced from a "Dummy-Tag" under the minecraft-namespace, with the same name. This results in modders only needing to create their own Tags under their own namespaces and at runtime just referencing the collected Tag, with all Tags other mods might have added. Data Pack authors would then only need to replace these collected Tags, in order to change game-behaviour.

### How to use it as a modder
Simply call some similiar lines of code before a Game is started (if you have a ForgeRegistry):
`ForgeTagManager.registerTagCollection(GameData.BLOCKS, ()-> ForgeTagCollection.fromForgeRegistry(ForgeRegistries.BLOCKS, ForgeTagCollection.getDefaultLoadingLocation(GameData.BLOCKS), "block"))`
Without a ForgeRegistry, here some code when faced by an Vanilla Registry:
`ForgeTagManager.registerTagCollection(FLUID_TAGS, ()-> new ForgeTagCollection<>(Fluid.REGISTRY, ForgeTagCollection.getDefaultLoadingLocation(ForgeTagManager.FLUID_TAGS), "fluid")`

Of course you can also use any custom Registry implementations you might have, as long as they can be boiled down to the necessary Functional Interfaces for Tag Loading. If it cannot (f.e. there aren't any Integer id's) you will need to handle loading and syncing on your own, in your own ForgeTagCollection subclass. 

### Things to be considered

- Extending the Tag Format, to allow for the condition System to be used with Tags (though maybe applying conditions for each sub-Tag?) (seperate PR)
- Make NetworkTagManager implement ISelectiveResourceReloadListener... (seperate PR)

MajorTuvok committed:
Initial Commit - revision of previous TagPR

MajorTuvok committed:
Merge branch '1.13-pre' of https://github.com/MinecraftForge/MinecraftForge into 1.13-registry-tags-v2

Removed some imports in Patches. I hope now that I noticed them, I found them all. As I already commited with some imports, more might be hard to find.

MajorTuvok committed:
Commit some patches that were somehow excluded and removed an import in the MinecraftServer patch.

williewillus commented:
Your section on thread safety are mostly incorrect, which leads me to believe most of the surrounding implementation you added for Forge is probably unnecessary.

Concurrent read-only access (iteration) to the same collection is thread-safe and does not constitute a data race. Moreover, in single player tags are not synced from the logical server to the logical client, the logical client directly accesses the same static fields.

williewillus commented:
also, how does this address my comments from last time?
In other words, what exactly does this provide for modders if they can just construct and manage a TagCollection on their own?

MajorTuvok commented:
@Williewillus sorry I pre-wrote that text and only noticed that they aren't applied from Server to Client in Singleplayer with the last commit...
As to the data race, thanks to pointing it out to me - I was missing the obvious fact that no one would modify these set's... (It's always the same with me isn't it?)

For your latter comment: Yes a modder can implement that on his own, but you now can with almost every possible Registry out there use the Forge System too. The better logging provided, the cross-mod compat and fewer effort that is needed is definitely useful in my opinion. If forge needs to extend Tags to all things, why not make it so that modders can use it too? 

Also the mapping from Tagged-Object to Tag should provide for faster access than the linear runtime cost used by vanilla - especially useful since Forge will almost certainly have more Tags than Vanilla. 

We should not forget that Immutable Tags will be faster, as most things will have a natural ordering defined or be IForgeRegistry entries.

MajorTuvok commented:
Note: I did strip all the utility Methods and generified quite a bit.

MajorTuvok committed:
Copy Vanilla TagWrapper Logic fully - would have to be done at some point. I sadly don't see a way to fetch the MinecraftServer to back TagWrapper's and therefore had to add all these stupid and annoying TagCollection fields. If anyone has a better idea (Vanilla approach really isn't nice and doesn't scale well) feel free to suggest it.

MajorTuvok committed:
Reduces "Log-Spam" by adding the enable Extended Logging flag to ForgeTagCollection, defaulting to false. May I add this as a config option (so that Data Pack authors can activate it as needed)?
Also fixed Enchantments using a plural Tag Name.

liach commented on patches/minecraft/net/minecraft/tags/Tag.java.patch:
```diff
@@ -0,0 +1,49 @@
+-      private final Set<Tag.ITagEntry<T>> field_200052_a = Sets.<Tag.ITagEntry<T>>newLinkedHashSet();
+-      private boolean field_200053_b;
++      protected final Set<Tag.ITagEntry<T>> field_200052_a = Sets.<Tag.ITagEntry<T>>newLinkedHashSet();
++      protected boolean field_200053_b;
```
no AT available?

MajorTuvok replied:
Changed

MajorTuvok commented:
As Immutable Tags can be expressed by creating a vanilla Tag approiatly, I 'll delete them and just make the TagBuilder return a Tag with similiar properties (as soon as I have time...)

MajorTuvok committed:
Merge branch '1.13-pre' of https://github.com/MinecraftForge/MinecraftForge into 1.13-registry-tags-v2

MajorTuvok committed:
Remove ImmutableTag and Formatting. I also did some testing and found out that sorting didn't occur as expected -> Sadly had to switch to testing Tagged Elements (@see ForgeTagBuilder).

liach commented on src/main/java/net/minecraftforge/common/tags/ForgeTagBuilder.java:
```diff
@@ -8,45 +12,123 @@
-            }
+    public static <T> Comparator<Tag<T>> tagIdComparator()
+    {
+        return Comparator.nullsFirst(Comparator.<Tag<T>, ResourceLocation>comparing(Tag::getId, Comparator.nullsFirst(Comparator.naturalOrder())));
```
Why do we need null first? If null happens, we should just crash.

MajorTuvok replied:
Ok

liach commented on src/main/java/net/minecraftforge/common/tags/ForgeTagBuilder.java:
```diff
@@ -173,19 +250,18 @@ public boolean resolve(Function<ResourceLocation, Tag<T>> resourceLocationToTag)
+    @FunctionalInterface
+    public interface TagFactory<T>
+    {
+        public Tag<T> build(ResourceLocation location, Set<ITagEntry<T>> entries, boolean preserveOrder, @Nullable Comparator<T> itemComparator);
```
try rename `item` to something else so that people don't confuse this with minecraft items.

liach commented on src/main/java/net/minecraftforge/common/tags/ForgeTagCollection.java:
```diff
@@ -95,13 +102,26 @@ public ForgeTagCollection(RegistryNamespaced<ResourceLocation, T> registryIn, Li
+    public ForgeTagCollection<T> comparingItemsBy(Comparator<T> comparator)
+    {
+        this.preserveOrder = false;
+        this.comparingItems = comparator;
```
same here, `item` -> `entry` (something more clear)

MajorTuvok committed:
Reduce patching by switching to AT's wherever possible. Also made the NetworkTagManager field a ForgeTagManager, in case someone needs it...

MajorTuvok commented on patches/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java.patch:
```diff
@@ -66,13 +66,6 @@
-    private void func_199814_a(NBTTagCompound p_199814_1_, Chunk p_199814_2_) {
-       NBTTagList nbttaglist = p_199814_1_.func_150295_c("Entities", 10);
--      World world = p_199814_2_.func_177412_p();
-+      World world = p_199814_2_.getWorld();
```
Where does this come from btw. (I didn't change anything there)? According to the generated decompiled files I have here this is still chunkIn.getWorld... (in the end this patch doesn't do anything...)?!?

williewillus replied:
simple: if a patch is generated and you didn't touch the file, don't commit the patch

MajorTuvok replied:
I didn't even notice it was in there...
-_-
(committed this by accident)

MajorTuvok committed:
Make provided Comparators fail on null values and document their behaviour. Also apply naming changes - originally these were named with *Item*, because MCP mappings for reasons I don't understand use this - But as someone else is of the same opinion like me (that using Item is confusing), I'm changing it here.

williewillus commented:
Include an example mod that clearly demonstrates how this is better than just using NetworkTagCollection on your own with minimal syncing work

MajorTuvok commented:
~~Ideally it should showcase using a ForgeRegistry and some kind of custom Registry, right?
I'll write a sample mod (that's what you meant with 'including' an example mod, right?) for this here and send the link as soon as possible.
(Though I'd first have o find out how to get a proper modder enviornment with this in it setup...)~~

liach commented on src/main/java/net/minecraftforge/common/tags/ForgeTagBuilder.java:
```diff
@@ -0,0 +1,281 @@
+     */
+    public static <T> Comparator<IForgeRegistryEntry<T>> registryNameComparator()
+    {
+        return Comparator.comparing(IForgeRegistryEntry::getRegistryName, Comparator.naturalOrder());
```
`Comparator.comparing(IForgeRegistryEntry::getRegistryName)` should be sufficient.

MajorTuvok replied:
I overlooked this overload...
Thanks

MajorTuvok committed:
Micro commit, shortening comparators even further

liach commented:
Vanilla don't sync function tags, I assume? Everything looks good.

MajorTuvok commented:
I'm [syncing Function Tags](https://github.com/MinecraftForge/MinecraftForge/pull/5289/files#diff-3b77432562c64460baf05c659c26fdd6R155) the same way Vanilla does...

LexManos commented on src/main/java/net/minecraftforge/common/tags/ForgeTagManager.java:
```diff
@@ -0,0 +1,299 @@
+    {
+        this.getBlocks().write(buffer);
+        this.getItems().write(buffer);
+        this.getFluids().write(buffer);
```
Welcome to the world of limited packet sizes. This will explode and explode fast. Do not hijack the vanilla packet.

MajorTuvok replied:
Ok, ~~so should I hook into the Forge Indexed Messages (I think that's how the Handschake Messages used to transferring Registry Data in 1.12 were called)?~~

LexManos commented on patches/minecraft/net/minecraft/tags/FluidTags.java.patch:
```diff
@@ -0,0 +1,18 @@
+ 
+-   public static class Wrapper extends Tag<Fluid> {
++   public static class Wrapper extends net.minecraftforge.common.tags.TagWrapper<Fluid>
++   {
```
Dont move the brace

MajorTuvok replied:
Sorry, ok

LexManos commented on patches/minecraft/net/minecraft/tags/NetworkTagCollection.java.patch:
```diff
@@ -0,0 +1,47 @@
++   protected final java.util.function.Function<T,Integer> itemToId;
++   protected final java.util.function.Function<Integer,T> idToItem;
+ 
++   public NetworkTagCollection(java.util.function.Predicate<ResourceLocation> isValueKnownPredicateIn, java.util.function.Function<ResourceLocation, T> resourceLocationToItemIn, java.util.function.Function<T, Integer> itemToId, java.util.function.Function<Integer, T> idToItem, String resourceLocationPrefixIn, boolean preserveOrderIn, String itemTypeNameIn) {
```
your parameter names are horrible, what is with the 'in' suffix? and 'getX' is a good format for getters...

MajorTuvok replied:
I just took over the mcp names I found elsewhere...
Will change it

LexManos replied:
MCP names are horrible crowd sourced things, pick real names that are appropriate.

LexManos commented on patches/minecraft/net/minecraft/tags/NetworkTagCollection.java.patch:
```diff
@@ -0,0 +1,47 @@
+    }
+ 
++   public void write(ResourceLocation collectionId, PacketBuffer buffer) { func_200042_a(buffer);} //called with this TagCollections ResourceLocation Id
++   public void read(ResourceLocation collectionId, PacketBuffer buffer) { func_200043_b(buffer);} //called with this TagCollections ResourceLocation Id
```
You take in a ID but never use it?

MajorTuvok replied:
Gr, I'm an idiot I know... (people can save that for themselves, if they need it for anything)

LexManos commented on patches/minecraft/net/minecraft/tags/NetworkTagManager.java.patch:
```diff
@@ -0,0 +1,28 @@
+-      networktagmanager.func_199717_a().func_200043_b(p_199714_0_);
+-      networktagmanager.func_199715_b().func_200043_b(p_199714_0_);
+-      networktagmanager.func_205704_c().func_200043_b(p_199714_0_);
++      net.minecraftforge.common.tags.ForgeTagManager networktagmanager = new net.minecraftforge.common.tags.ForgeTagManager();
```
Do not hijack the vanilla packets.

MajorTuvok replied:
...
~~FMLIndexedMessage (or whatever it is called) as an alternative?~~

LexManos replied:
Forge has a handshake system, and a bunch of other things are are sent during play time. Sending tags as seperate lists avoids the single massive packet issue. And should be fine as they are just swapped out in memory.
Basically stop tieing everything together as it'll easily explode the limit that S->C packets have.

MajorTuvok commented:
I'll switch out the Function Tags Tag collection with an ForgeTagCollection, if that doesn't require too much patching...?

LexManos commented on patches/minecraft/net/minecraft/tags/NetworkTagManager.java.patch:
```diff
@@ -0,0 +1,28 @@
+-   private final NetworkTagCollection<Fluid> field_205705_c = new NetworkTagCollection<Fluid>(Fluid.field_205588_a, "tags/fluids", "fluid");
++   private final NetworkTagCollection<Block> field_199719_a = net.minecraftforge.common.tags.ForgeTagCollection.fromForgeRegistry(net.minecraftforge.registries.ForgeRegistries.BLOCKS, net.minecraftforge.common.tags.ForgeTagCollection.getDefaultLoadingLocation(net.minecraftforge.registries.GameData.BLOCKS),"block");
++   private final NetworkTagCollection<Item> field_199720_b = net.minecraftforge.common.tags.ForgeTagCollection.fromForgeRegistry(net.minecraftforge.registries.ForgeRegistries.ITEMS, net.minecraftforge.common.tags.ForgeTagCollection.getDefaultLoadingLocation(net.minecraftforge.registries.GameData.ITEMS),"item");
++   private final NetworkTagCollection<Fluid> field_205705_c = new net.minecraftforge.common.tags.ForgeTagCollection<>(Fluid.field_205588_a, net.minecraftforge.common.tags.ForgeTagCollection.getDefaultLoadingLocation(net.minecraftforge.common.tags.ForgeTagManager.FLUID_TAGS), "fluid");
```
Why do we even need to replace the vanilla tags?

MajorTuvok replied:
My reasoning was to provide them with the better loading...
Though when moving towards AT, I already thought, that it might be better to just let these things exist and then return our own TagCollections in the getters?

MajorTuvok commented:
Ok, so I just remembered why I didn't switch over to these handshake Messages:
reloading. 
Using the handshake Messages, Tags would only be synced when the connection was initiated, and never thereafter (including not when Tags are reloaded...).

Using the Vanilla Packet seemed a pretty good choice to me, when I noticed it (It isn't as Lex reasoned out above...).

So my next idea would be to send/read the vanilla Packet as is, when connected to an vanilla client/server, but send one of these Packets for each Tag Collection seperatly when using a Forge/Forge connection. Any thoughts on this abusal of the Tag Update Packet? (declaring a seperate Packet isn't an option)

MajorTuvok commented:
Additionally, if you guys don't mind, I'll make ForgeTagManager a singleton, which will save a lot of code... (And make the registration of TagCollections less complex on the inside)

williewillus commented:
why is making a separate packet not an option? vanilla clients connecting to a forge server won't be able to make use of any extra tags anyway, if they can even connect in this scenario.

MajorTuvok commented:
The Packet would need go be registered, even when being connected to vanilla... Only because Forge suddenly has some Entity Tags registered, it shouldn't prevent connection to vanilla (Forge would loose vanilla compat then) 

Hm, though I think you're right @williewillus,  as long as the Packed gains an id above all the vanilla packets, it should not interfere with them... Does Forge do this already somewhere?

MajorTuvok committed:
Merge branch '1.13-pre' of https://github.com/MinecraftForge/MinecraftForge into 1.13-registry-tags-v2

MajorTuvok committed:
Apply a few suggestions and make ForgeTagManager a singleton.

I also merged in the current 1.13-pre branch, which has the strange side-effect, that every time I run setup now, it no longer recognizes my changes to the forge_at file?!? I hope this doesn't cause any Problems with this commit.
Additionally I tried to remove this strange re-add things patch, let's see whether it worked or not.

MajorTuvok committed:
This should remove the strange AnvilChunkLoader modification

MajorTuvok committed:
For reasons I don't understand, IntelliJ seems to allow only one whitespace change at the time - Part1

MajorTuvok committed:
For reasons I don't understand, IntelliJ seems to allow only one whitespace change at the time - Part2

MajorTuvok committed:
For reasons I don't understand, IntelliJ seems to allow only one whitespace change at the time - Part3

MajorTuvok commented:
Grr, sometimes IntelliJ can be a bit annoying - It always removed one of the 2 whitespaces, as soon as I opened the commit window - I honly noticed that when I was committing the Part2 thing... 
So this dammned thing took me 3 commits!

MajorTuvok committed:
Merge branch '1.13-pre' of https://github.com/MinecraftForge/MinecraftForge into 1.13-registry-tags-v2

MajorTuvok committed:
Properly handle

MajorTuvok committed:
Properly Handle Function Tags... Revert accidentally committed Files

MajorTuvok committed:
Go away AnvilChunkLoader!!!
